### DISCUSSION TITLE
Incorrect sliderball rendering with legacy skin

### DISCUSSION THREAD

**Discussion by sineplusx** [ppy/osu#21355] [Q&A] [answered]
### Type

Cosmetic

### Bug description

While testing my skin for lazer, I noticed that the sliderball is rendered/drawn incorrectly in a special occasion.

In my skin, I basically implemented the followcircle into the `sliderb` texture so that it wont grow when going over the score points. I also have the sliderball tinted in combo colors, but I wanted the color to be brighter. I duplicated the `sliderb` texture and made a `sliderb-spec` texture with the same image but at low transparency.

`sliderb`
![sliderb@2x](https://user-images.githubusercontent.com/112003827/202013654-6da82f48-79cf-4148-91b6-4edc12020015.png)
`sliderb-spec`
![sliderb-spec@2x](https://user-images.githubusercontent.com/112003827/202013695-c14e9fb8-38bf-4f66-b7bf-5af4c74c67ca.png)

In summary, the `sliderb` texture will be multiplied with the combo color and the `sliderb-spec` texture should be perfectly layered on top. And this is not the case. When starting a slider, the textures will be perfectly stacked, but as the slider goes on, they will separate in position, enough to be noticeable and annoying. I have put screenshots and a video in the section below.

Maybe related?: #4288

Aside from that, while researching the issue: Both the [osu! wiki](https://osu.ppy.sh/wiki/en/Skinning/osu%21#slider) and the community-maintained [skinnable files list](https://docs.google.com/spreadsheets/d/1bhnV-CQRMy3Z0npQd9XSoTdkYxz0ew5e648S00qkJZ8/edit#gid=2074725196&range=C178) state that the `sliderb-spec` texture will not be drawn if a skinned `sliderb` texture is present. I haven't tested this in stable, but if it is true, then lazer is handling the sliderball draw process differently than stable.

Used Skin: [skin.zip](https://github.com/ppy/osu/files/10015834/skin.zip)

### Screenshots or videos

![stacked](https://user-images.githubusercontent.com/112003827/202014973-1afa28d9-a3c5-4d32-825a-4430d5411903.png)
![bug](https://user-images.githubusercontent.com/112003827/202015012-42018d55-88ef-492b-b734-f7cd3e055867.png)
https://user-images.githubusercontent.com/112003827/202015249-cea594c5-e94d-4dc1-a7f4-5f1f231764b9.mp4

### Version

2022.1101.0

### Logs

(Will add Logs if needed)

**Comment by peppy** [comment]
Based on the video alone, this doesn't look like a frame delay as you suspected, but an issue with rotation being applied to one of the two elements. You can see in the first slider in the video that everything looks okay until the slider turns around the bend.

Edit: confirmed by rotating the skin elements as a test


https://user-images.githubusercontent.com/191335/202982553-deebca46-9376-4309-8681-62cc16ee2ae7.mp4

That said, it still doesn't match stable since you're not seeing it there, so I'll investigate further.

**Reply by sineplusx** [reply]
Thanks for the help. I should have really looked into that possibility before opening an issue.

**Comment by peppy** [comment]
Okay, so basically the issue here is that the skin has incorrectly centered textures, and you haven't seen this on stable because `sliderb-spec@2x.png` is hidden per spec.

So the only required fix here is to not show `spiderb-spec` and `sliderb-nd` when `sliderb` is present and skinned.

**Comment by peppy** [comment, marked as answer]
I'm not sure this should be fixed. Fixing is not simple because the original logic seems broken. Basically skinners could only replace `-nd` and `-spec` if they lived with the default skin's "stripes" still existing. It's probably better than we make a breaking change here and allow for all three to be customised however a skinner wants to.

The fix for your skin would be to delete `sliderb-spec`.

I'll convert this to a discussion and let's see if anyone else hits the issue.

**Reply by peppy** [reply]
@Walavouchey I'm not sure where to track this but looping you in because it may require wiki changes. I don't mind changing stable's behaviour to match if that makes this easier.

**Reply by Walavouchey** [reply]
ppy/osu-stable-issues, hl me and @ RockRoller01 (maintains the spreadsheet linked in op)

**Reply by sineplusx** [reply]
The fix for my skin would be fixing the centering issue in both textures. After all, I intended to use the `spec` texture to make the brighter color

**Reply by peppy** [reply]
Just keep in mind the `spec` layer will not show in stable unless we make the change.
